### PR TITLE
periph/wdg: independent watchdog timer support on stm32l1 + common .h file

### PR DESCRIPTION
**Pull request by Cr0s**
Independent Watchdog timer (`IWDG`) support for `L1` family and (maybe) a common interface for watchdog implementations on other `STM32` families or other MCUs with watchdogs working by the same principle: prescaler & reload value.

Cheers!

### CODE REVIEW

**Comment by haukepetersen** [comment]
Thanks for your contribution, I think it is very useful! 

Regarding the general interface, I think it is a little bit to fixed on the STMs. As the periph itnerfaces aim to be as vendor (and architecture) independent and simple as possible, I think we need to simplify and or generalize the interface a little bit. I would rather propose something in the direction of
```
int wdt_enable(uint32_t timeout);
void wdt_reload(void);
int wdt_cased_reset(void);
```
All the prescaler handling etc should be handled by a CPUs implementation (e.g. by returning an error code if the given timeout can not be configured).

So for this PR I would propose one of the following options: a) we try to confine the perih/wdt interface together and adapt the implementation for the STMs (supposedly for all of them at once in `stm32_common/periph/wdt.c`) or b) you move your proposed iwdt interface into the stm32l1 include folder for now and we re-factor it later on the go. What do you say?

File: drivers/include/periph/wdg.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2016 Unwired Devices [info@unwds.com]
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup
```
Comment (OlegHahm) [inline drivers/include/periph/wdg.h]: The group should have a name and description.

File: drivers/include/periph/wdg.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ * @ingroup
+ * @brief
+ * @{
+ * @file		wdg.h
```
Comment (OlegHahm) [inline drivers/include/periph/wdg.h]: remove the parameter, doxygen will figure it out itself.

File: drivers/include/periph/wdg.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ */
+#ifndef DRIVERS_INCLUDE_PERIPH_WDG_H_
+#define DRIVERS_INCLUDE_PERIPH_WDG_H_
+
+#include <stdbool.h>
+
+/**
+ * @brief Sets watchdog timer prescaler.
+ *
+ * @param prescaler watchdog timer prescaler value
```
Comment (OlegHahm) [inline drivers/include/periph/wdg.h]: Please add `[in]` or `[out]` for all the function parameters to specify if they are input or output parameters.

File: drivers/include/periph/wdg.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ */
+void wdg_set_prescaler(uint8_t prescaler);
+
+/**
+ * @brief Sets watchdog timer reload value.
+ *
+ * @param reload watchdog timer reload value
```
Comment (OlegHahm) [inline drivers/include/periph/wdg.h]: I think we should specify the unit for the parameter

**Comment by aabadie** [comment]
ping @Cr0s

**Comment by Cr0s** [comment]
Will address the issues ASAP.

File: cpu/stm32l1/periph/iwdg.c [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup
+ * @ingroup
+ * @brief
+ * @{
+ * @file		iwdg.com
```
Comment (kaspar030) [inline cpu/stm32l1/periph/iwdg.c]: filename seems wrong

File: cpu/stm32l1/periph/iwdg.c [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+ * @brief       STM32L1 Independent Watchdog
+ * @author      EP [ep@unwds.com]
```
Comment (kaspar030) [inline cpu/stm32l1/periph/iwdg.c]: please add real name

File: cpu/stm32l1/periph/iwdg.c [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+static inline void iwdg_lock(void)
+{
+    IWDG->KR = IWDG_LOCK;
+}
+
+void wdg_set_prescaler(uint8_t prescaler)
+{
+    /* Check valid prescaler values */
+    assert(prescaler == 0x00    ||
```
Comment (kaspar030) [inline cpu/stm32l1/periph/iwdg.c]: ```assert(prescaler <= 0x6)```?

File: tests/periph_wdg/Makefile [hunk 8]
```diff
@@ -0,0 +1,6 @@
+export APPLICATION = periph_wdg
```
Comment (kaspar030) [inline tests/periph_wdg/Makefile]: drop export

File: tests/periph_wdg/main.c [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+#include "cpu_conf.h"
+#include "periph/wdg.h"
+#include "xtimer.h"
+
+#define WDG_RELOAD (uint16_t) 0x0FFF
+#define WDG_PRESCALER 0x03
+
+int main(void)
+{
+    xtimer_init();
```
Comment (kaspar030) [inline tests/periph_wdg/main.c]: AFAIR this is called by auto_init, drop it here

File: tests/periph_wdg/main.c [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    wdg_set_reload(WDG_RELOAD);
+    wdg_reload();
+
+	/* Enable WDG */
+    wdg_enable();
+
+    puts("[!] Watchdog enabled. Sleeping 150 ms...");
+
+    /* Wait 150 ms */
+    xtimer_usleep(1000 * 150);
```
Comment (kaspar030) [inline tests/periph_wdg/main.c]: 1000LU

File: tests/periph_wdg/main.c [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+    puts("[!] Reloading watchdog...");
+    wdg_reload();
+
+    /* Freeze */
+    puts("[!] Watchdog reloaded. Now freezing...");
+
+    volatile int seconds;
+    for (seconds = 0;; seconds++) {
```
Comment (kaspar030) [inline tests/periph_wdg/main.c]: please move initialization into for(). why volatile?

File: tests/periph_wdg/main.c [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    xtimer_sleep(5);
+
+    puts("Test for the WDG driver");
+    puts("This test will setup the WDG, reload it and then hangs up in infinite loop\n");
+
+    /* Check that previous reset was initiated by watchdog timer  */
+    if (wdg_reset_occurred()) {
+        puts("[!] Reset by WDG occurred");
+
+        for (;; ) ;  /* Hang here */
```
Comment (kaspar030) [inline tests/periph_wdg/main.c]: please use ```while (1) { }```

**Comment by tcschmidt** [comment]
@Cr0s any plans to address the requests?

**Comment by tcschmidt** [comment]
Closed due to long-term inactivity.

**Pull request closed without merging**